e2e/rbd.go

```diff
@@ -55,104 +55,92 @@ var (
-			e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+csiDriverObject, err)
-		}
-	} else {
-		_, err = framework.RunKubectlInput(cephCSINamespace, string(csiDriver), action, "-f", "-")
+func deployRBDPlugin() {
+	// create a rbd deployer
+	rd := rbdDeploy{deployer: deploy{}, templates: rbdTemplates}
+
+	// delete objects deployed by rook
+	deleteRookTemplates := []string{rbdDirPath + rbdProvisionerRBAC, rbdDirPath + rbdNodePluginRBAC}
+	for _, template := range deleteRookTemplates {
+		err := rd.runAction(template, "delete", ns)
 		if err != nil {
-			e2elog.Failf("failed to %s CSIDriver object with error %v", action, err)
+			e2elog.Failf("failed to delete rook template with error %v", err)
 		}
 	}
-	data, err := replaceNamespaceInTemplate(rbdDirPath + rbdProvisioner)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisioner, err)
-	}
-	data = oneReplicaDeployYaml(data)
-	data = enableTopologyInTemplate(data)
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s rbd provisioner with error %v", action, err)
-	}
-
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdProvisionerRBAC)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisionerRBAC, err)
-	}
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s provisioner rbac with error %v", action, err)
-	}
-
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdProvisionerPSP)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisionerPSP, err)
-	}
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s provisioner psp with error %v", action, err)
-	}
-
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdNodePlugin)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdNodePlugin, err)
-	}
 
-	domainLabel := nodeRegionLabel + "," + nodeZoneLabel
-	data = addTopologyDomainsToDSYaml(data, domainLabel)
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s nodeplugin with error %v", action, err)
+	// Provision RBD from the CSI templates
+	for _, template := range rd.templates {
+		rd.changeTemplateAndRunAction(template, "create")
 	}
+}
 
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdNodePluginRBAC)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdNodePluginRBAC, err)
-	}
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s nodeplugin rbac with error %v", action, err)
+func (d rbdDeploy) changeTemplateAndRunAction(template, action string) {
```

Comment:
return error from `changeTemplateAndRunAction` let the caller handle the error message?

Reply:
sure. done that way too.  Thanks @Madhu-1